auth_mellon_util.c

```diff
@@ -1507,32 +1507,104 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
+ */
+
+char *
+am_strip_whitespace_and_quotes(char *str)
+{
+    char *begin, *end;
+
+    /* Skip leading whitespace */
+    for (begin = str; *begin && isspace(*begin); begin++);
+
+    /* Find end of string */
+    for (end = begin; *end; end++);
+
+    /* Skip trailing whitespace */
+    for (end = end > begin ? end-1 : end; end > begin && isspace(*end); end--);
+
+    /* Nuke beginning & ending quotes */
+    if (end > begin && *begin == '"' && *end == '"') {
+        begin++;
+        end--;
+    }
+
+    /* Null terminate */
+    if (end > begin) {
+        end[1] = 0;
+    }
+
+    return begin;
+}
+
+/* This function checks if an HTTP PAOS header is valid and
+ * returns any service options which may have been specified.
+ *
+ * A PAOS header is composed of a mandatory PAOS version and service
+ * values. A semicolon separates the version from the service values.
+ *
+ * Service values are delimited by semicolons, and options are
+ * comma-delimited from the service value and each other.
+ *
+ * The PAOS version must be in the form ver="xxx" (note the version
+ * string must be in double quotes).
+ *
+ * The ECP service must be specified, it MAY be followed by optional
+ * comma seperated options, all values must be in double quotes.
+ *
+ * ECP Service
+ *   "urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp"
+ *
+ * Recognized Options:
+ *
+ * Support for channel bindings
+ *  urn:oasis:names:tc:SAML:protocol:ext:channel-binding
+ *
+ * Support for Holder-of-Key subject confirmation
+ *   urn:oasis:names:tc:SAML:2.0:cm:holder-of-key
+ *
+ * Request for signed SAML request
+ *   urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp:2.0:WantAuthnRequestsSigned
+ *
+ * Request to delegate credentials to the service provider
+ *   urn:oasis:names:tc:SAML:2.0:conditions:delegation
+ *
+ *
+ * Example PAOS HTP header::
+ *
+ *   PAOS: ver="urn:liberty:paos:2003-08";
+ *     "urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp",
+ *     "urn:oasis:names:tc:SAML:protocol:ext:channel-binding",
+ *     "urn:oasis:names:tc:SAML:2.0:cm:holder-of-key"
  *
  * Parameters:
- *  request_rec *r         The request
- *  const char *header     The PAOS header value
+ *  request_rec *r              The request
+ *  const char *header          The PAOS header value
+ *  ECPServiceOptions *options_return
+ *                              Pointer to location to receive options,
+ *                              may be NULL. Bitmask of option flags.
  *
  * Returns:
- *   true if the PAOS header is valid, false otherwise
+ *   true if the PAOS header is valid, false otherwise. If options is non-NULL
+ *   then the set of option flags is returned there.
  *
  */
-#define PAOS_VERSION_TOKEN "ver=\"" LASSO_PAOS_HREF "\""
-#define ECP_SERVICE_TOKEN  "\"" LASSO_ECP_HREF "\""
-bool am_validate_paos_header(request_rec *r, const char *header)
+bool am_validate_paos_header(request_rec *r, const char *header,
+                             ECPServiceOptions *options_return)
```

Comment:
I wonder if it may be time to rename the function to something like `am_parse_paos_header`? It is in effect parsing it, and returning false if it is invalid.